dns: introduce ResolutionStatus for ResolveCb and fix #9927

junr03 opened the pull request:
Description: this PR introduces ResolutionStatus to the ResolveCb. This status can be used to determine if DnsResolution was successful or not, and lets the callback target adjust accordingly. It is used to fix #9927. Additionally, it can be used to fix
https://github.com/lyft/envoy-mobile/issues/673 in a subsequent PR.
Risk Level: med. This PR changes the behavior of Strict Dns Cluster. The introduced argument is not used anywhere, so all behavior stays the same.
Testing: Updated tests to pass the new argument to the callback. But behavior stays the same.
Docs Changes: updated config docs and discovery docs.
Release Notes: added a release note.

Fixes #9927 

Signed-off-by: Jose Nino <jnino@lyft.com>

junr03 commented:
@mattklein123 I finally got time to come back to this. I started by just introducing the new argument for the callback, but leaving all behavior unchanged. I didn't want to conflate this with the two independent changes (don't update hosts on failure, allow different timeouts on failure) needed for the linked issues.

Let me know if you'd rather look at it all together.

junr03 commented on include/envoy/network/dns.h:
```diff
@@ -45,12 +45,17 @@ class DnsResolver {
+
   /**
    * Called when a resolution attempt is complete.
-   * @param response supplies the list of resolved IP addresses and TTLs. The list will be empty if
```
> The list will be empty if the resolution failed.

Most of the callback targets actually use this implicit "failure" hint. The strict dns cluster does not, which is why #9927 happens.

I am perusing through c-ares and trying to assert if according to RFC we can reasonably expect ARES_SUCCESS with no resolved addresses. @mattklein123 do you know off of the top of your mind?

If that is the case, then we don't _necessarily_ need to add the status. Just make sure the callback targets behave as we expect when the address list is empty.

mattklein123 replied:
Yes it is definitely possible for a DNS resolution to return nothing and be successful, so we definitely need to handle this differently. What we do today is really just a hack. If higher layer code *additionally* wants to stabilize hosts in the case of 0 legitimate responses that is fine (e.g. I think logical DNS should do this, but strict DNS should not).

junr03 replied:
Yep, agreed it is a hack. But I wanted to assert if it was a hack based on actual possible behavior; c-ares would only have an empty list with non-ARES_SUCCESS status. It wasn't immediately clear to me from my reading. Given that is not the case, I'll go ahead and continue with this change. Thanks!

mattklein123 commented on source/common/network/dns_impl.cc:
```diff
@@ -89,7 +89,7 @@ void DnsResolverImpl::PendingResolution::onAresGetAddrInfoCallback(int status, i
     // can be done with this query and initiate a new one.
     if (!cancelled_) {
-      callback_({});
+      callback_(ResolutionStatus::Failure, {});
```
Should you have another status like cancelled/destructing/etc.? This is not quite a failure? I'm not sure it matters to higher layer code but it might?

junr03 replied:
I thought about this when I first introduced the destruction behavior. My first opinion is that higher layers only pivot behavior based on success/failure so I was going to keep it at that level of complexity for this change, and the subsequent PR where I fix #9927. 

Then I was going to explore the need to add a `Destruction` value that allow us to set tighter timers than on "regular failure" when fixing for https://github.com/lyft/envoy-mobile/issues/673. But I am not sure if we need different behavior based on what particular failure scenario the resolver experienced. By the same token, people might want even more granularity in the future?

In short: I did not want to commit to more granular status in this initial PR, and I don't think it is going to be that bad to add more granularity, if desired, in subsequent PRs.

mattklein123 commented on test/common/network/dns_impl_test.cc:
```diff
@@ -480,11 +480,12 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, DnsImplTest,
-                                             });
+  ActiveDnsQuery* query = resolver_->resolve(
+      "", DnsLookupFamily::V4Only,
+      [&](DnsResolver::ResolutionStatus, std::list<DnsResponse>&& results) -> void {
```
Can you actually add expectations around resolution status in this file where appropriate?

junr03 replied:
Yep! I didn't want to spend too much time creating all of this until we discussed https://github.com/envoyproxy/envoy/pull/10137#discussion_r382861884. Will update

mattklein123 commented on test/common/upstream/logical_dns_cluster_test.cc:
```diff
@@ -156,10 +159,12 @@ class LogicalDnsClusterTest : public testing::Test {
-    // Empty should not cause any change.
+    // Failure should not cause any change.
+    // TODO(junr03): make sure to have coverage here. Why does strict change but logical not? Why
+    // this inconsistency? Updated in subsequent change.
```
See my other comment. IMO logical DNS should "stabilize" but strict should not in the case of a successful nothing response, but let's make sure to clearly document this.

junr03 replied:
chatted about this offline, as I was thinking that the suggested behavior change was a little odd. I am now clear on the expectation.

We also agreed that at this point we can just do #9927 here. I will update this today.

junr03 commented:
@mattklein123 updated. I have updated dns_impl_test for expectations around status in different conditions. I also updated higher level constructs to use status, but I _didn't_ change any behavior yet based on this new information (other than adding more relevant stats where possible). I plan on addressing https://github.com/envoyproxy/envoy/pull/10137#discussion_r383415588 and hence #9927 in a subsequent PR. Let me know if you'd rather see it here.

junr03 commented on source/common/upstream/logical_dns_cluster.cc:
```diff
@@ -100,23 +100,24 @@ void LogicalDnsCluster::startResolve() {
               Network::Utility::getAddressWithPort(*(response.front().address_),
                                                    Network::Utility::portFromTcpUrl(dns_url_));
 
-          if (respect_dns_ttl_ && response.front().ttl_ != std::chrono::seconds(0)) {
```
moved this for consistency with the strict dns cluster code. But happy to move back up.

junr03 commented on test/common/upstream/logical_dns_cluster_test.cc:
```diff
@@ -272,6 +288,88 @@ TEST_P(LogicalDnsParamTest, ImmediateResolve) {
+  tls_.shutdownThread();
+}
+
+TEST_F(LogicalDnsParamTest, TtlAsDnsRefreshRate) {
```
Adding coverage that did not exist.

mattklein123 commented on test/common/network/dns_impl_test.cc:
```diff
@@ -706,7 +742,10 @@ TEST_P(DnsImplTest, RemoteAsyncLookup) {
-                                        [&](std::list<DnsResponse>&& results) -> void {
+                                        [&](DnsResolver::ResolutionStatus status,
+                                            std::list<DnsResponse>&& results) -> void {
+                                          // Failure f
```
?

mattklein123 commented on test/common/network/dns_impl_test.cc:
```diff
@@ -715,7 +754,8 @@ TEST_P(DnsImplTest, RemoteAsyncLookup) {
   EXPECT_NE(nullptr, resolver_->resolve("some.good.domain", DnsLookupFamily::Auto,
-                                        [&](std::list<DnsResponse>&& results) -> void {
+                                        [&](DnsResolver::ResolutionStatus,
+                                            std::list<DnsResponse>&& results) -> void {
```
I'm not sure what the best way would be but is there some way to simplify all of thee calls to make sure have a success/failure expectation on all of them? Perhaps some helper function/object that does the resolve, does the expectations, and stores the resolution list?

junr03 replied:
folded everything into `resolveWithExpectations`

junr03 commented on test/common/network/dns_impl_test.cc:
```diff
@@ -635,279 +649,174 @@ TEST_P(DnsImplTest, DestroyChannelOnRefused) {
-  EXPECT_TRUE(hasAddress(address_list, "201.134.56.7"));
-}
-
-TEST_P(DnsImplTest, DnsIpAddressVersion) {
```
moved above

mattklein123 commented on test/common/network/dns_impl_test.cc:
```diff
@@ -495,117 +523,107 @@ TEST_P(DnsImplTest, DestructPending) {
-                         [&](const std::list<DnsResponse> &
-                             /*results*/) -> void { throw EnvoyException("Envoy exception"); }));
+  EXPECT_EQ(nullptr,
+            resolver_->resolve(
```
qq: Can't we use the new resolve with expectations everywhere?

junr03 replied:
yes! I thought it might be overkill given that there were far fewer call sites. But if I'm cleaning up something why not go the entire way :). Now everything (for real) is folded into helper functions.

junr03 commented:
Looks like the circle machines have a funny hosts file?

```
[----------] 1 test from IpVersions/DnsImplTest
[ RUN      ] IpVersions/DnsImplTest.LocalLookup/IPv4
test/common/network/dns_impl_test.cc:470: Failure
Expected equality of these values:
  expected_results
    Which is: { "127.0.0.1" }
  address_as_string_list
    Which is: { "127.0.0.1", "127.0.0.1" }
```

mattklein123 commented:
Thanks for the fix. If that's the CI error you are seeing it seems like it must have to do with your change?

junr03 commented:
Yep, definitely has to do with my change. I just wanted to make sure that I was surfacing something latent in circle ci and not an unintended regression. Previously the `LocalLookup` test only asserted that `127.0.0.1` was contained in the resolved address list, not exact match like now. I opened https://github.com/envoyproxy/envoy/pull/10197 to verify. It looks like our circle CI setup ends up with a hosts file with a double entry for localhost. https://app.circleci.com/jobs/github/envoyproxy/envoy/324116.
